content/blog/kallisti5/2018-05-19_rune_-_haiku_images_on_arm.md

```diff
@@ -0,0 +1,65 @@
++++
+type = "blog"
+title = "Rune - Haiku Images on ARM"
+author = "kallisti5"
+date = "2018-05-19 15:33:42-05:00"
+tags = ["haiku", "software"]
++++
+
+Up until recently, Haiku builds for ARM have targetted individual ARM boards. The compile process for ARM images required two things: an architecture, and a target board (such as the Raspberry Pi 2). This board setting adjusted a large number of defines throughout Haiku at compile time to set the operating system up for the target ARM device. The board selection also handled placing all the propriety bits (a lot of which have sketchy licensing) into the Haiku image during compile. Haiku would then has to distribute these files. (sketchy licensing and all)
+
+Over the past few years, François Revol, Ithamar R. Adema, and others have worked to add [Flat Device Tree (FDT)](https://elinux.org/Device_Tree_Reference) support to Haiku. FDT's enable operating systems to obtain core knowledge of the devices they run on by simply swapping one or more compiled binary files. These files describe critical things the operating system needs to know about the hardware they run on. *Really* important things such as what devices exist at what memory locations. (Think video frame buffers, serial ports, etc)
+
+In a series of cryptic commits in July 2017, I removed these board-centric build steps with grand plans of making testing (and running) Haiku on ARM devices *easier*.
+
+# Strange Inspirations
+
+Inspiration comes from strange places. In 2016, I started playing around with [Fedora on ARM](https://arm.fedoraproject.org/).  Fedora instructs you to download a generic ARM Image. With my experience and pains with Haiku on ARM, this seemed odd to me.  How can they give you "one image" which works on all devices? There are a huge number of properitery bits which you need to include in boot partitions for each board... some ARM boards even require binary blobs copied to special offset addresses outside of the filesystem. Fedora's answer (and magic) lies in their "arm-image-installer" script.
+
+Fedora's arm-image-installer tool:
+
+  * Takes the Fedora ARM image you download
+  * You specify the target board (Like the Raspberry Pi 3)
+  * Takes the properiety bootloader artifacts from your local system
+  * Selects the correct FDT file
+  * Burns all of this to an SD card ready to boot on your device
+
+So, end users use the "arm-image-installer" tool to "burn" the ARM image to an SD card (and doing a bunch of magic under the covers)
+
+{{< alert-warning "Spoiler" "Haiku on ARM is in its infancy. Don't expect any magic. 'rpi2' will likely work the 'best' which involves loading out kernel and spitting out information to the serial uart.">}}
```

Comment:
Don't we have framebuffer support on the Beagle xM? That would be a little more exciting :D

Reply:
We do.. just need to get the FDT's in order (as well as some u-boot binaries)